### DISCUSSION TITLE
Implement multiple dispatch

### DISCUSSION THREAD

**Discussion by EmilyBourne** [pyccel/pyccel#1334] [Ideas]
**Describe the bug**
Sometimes we want different functions to run different methods depending on the type. In python this is often handled with the `isinstance` decorator, however in pyccel this would be unmanageable. We would therefore prefer to use multiple dispatch. Implementation of this would fix issue #802 

**Test Code**
Provide code which does not currently work but which should do when this issue is fixed:
```python
@dispatch('float')
def get_real(a : 'float'):
    return a

@dispatch('complex')
def get_real(a : 'complex'):
    return a.real
```
The dispatch decorator can either be written ourselves in pyccel or we could add a dependency to [multipledispatch](https://pypi.org/project/multipledispatch/) (last pypi update Jun 2023)

This code should be translated to an `Interface`. In C and fortran this should work out of the box, in python we need to print the individual functions with the dispatch decorator when necessary. Ideally these functions should be grouped if the code is identical. E.g:
```python
@dispatch(('int'),('float'))
@template('T',['int','float'])
def get_real(a : 'T'):
    return a

@dispatch('complex')
def get_real(a : 'complex'):
    return a.real
```

**Comment by EmilyBourne** [comment]
Blocked by #886

**Comment by EmilyBourne** [comment]
@yguclu This issue was opened following a discussion with you and @ratnania . Why has it been converted to a discussion?

**Reply by yguclu** [reply]
I am currently listing Pyccel projects as discussions, as they will need a discussion among developers before the actual implementation.

**Reply by EmilyBourne** [reply]
Ok. Which points of this issue do you think need clarifying by a discussion? The choice of the python implementation of `@dispatch`?

**Reply by yguclu** [reply]
Of course we could implement this in pure Python, following the `multipledispatch` library. Faster implementations would have to be done in Fortran and C. If we decide to do this, we have to discuss how the generated code should look like.

**Reply by EmilyBourne** [reply]
I'm not sure I see what needs discussing? The generated code is the easy case as we already have `Interface`s in Fortran and C. This already requires us to print a different function for each combination of argument types. The case where the content is different does not require any special treatment

**Reply by yguclu** [reply]
You are right, most of the machinery is already implemented and is used to handle different argument types for a single function.

I still think that a few things have not been clarified yet. For example, do we need to provide arguments to the `@dispatch` decorator if we use type annotations (or the `@type` decorator)? This code duplication looks confusing to me, and error prone. Further, as you mentioned, the Python implementation should be decided.

**Comment by EmilyBourne** [comment]
An alternative to the `multipledispatch` library would be the Python built-in [`functools.singledispatch`](https://docs.python.org/3/library/functools.html#functools.singledispatch)

It works like this:
```python
from functools import singledispatch


@singledispatch
def func(val):
    raise NotImplementedError


@func.register
def _(val: str):
    print('This is a string')


@func.register
def _(val: int):
    print('This is an int')
```

This seems like something that would be easy to print, although it requires slightly more effort to parse. However I'm not sure how NumPy types should be managed

**Comment by said-hadjout** [comment]
We can use the `@overload` decorator provided by the typing module https://docs.python.org/3/library/typing.html#overload.
I don't think we need to worry about the Python implementation, we can use it as a first step internally for the external libraries, and decide afterward if we want to expose it to the user

**Reply by EmilyBourne** [reply]
I agree that the typing module could be helpful however I disagree about the Python implementation. It is often frustrating to have to deal with Python as a special case when it should be trivial.

I have had a look at the typing module. At first I thought it took care of the dispatch but instead it seems closer to `@template` in that it lets you group combinations of arguments. I therefore don't think that this is the right solution as in Pyccel we always ensure that our implementations reproduce the same behaviour as in Python. I don't think this would be the case with an implementation using `@overload` unless I've missed something?

**Reply by said-hadjout** [reply]
The Python implementation is needed if we need to expose it to the user, for the moment we can just create our own decorator that does nothing, maybe this way is better

**Reply by EmilyBourne** [reply]
> The Python implementation is needed if we need to expose it to the user

It is also needed for the tests (which was the original reason for opening this issue)

**Reply by said-hadjout** [reply]
Yes, it is preferable for it to work in Python, but if we can't decide about the Python behavior, we can postpone it until it is clear, because regardless of the behavior we want to give it in Python, the implementation in Fortran/C will be the same

**Reply by EmilyBourne** [reply]
I agree that the implementation in Fortran/C won't change. But I don't think it should be difficult to reach a consensus on a method that would work for Python so I don't see any reason to write an implementation that we plan to replace when we can just think a little longer and get it right first time.

**Comment by EmilyBourne** [comment]
I do think we should try and find a solution to this though. This issue comes back repeatedly. Currently we have 4 suggestions:
1. `@dispatch` from [multipledispatch](https://pypi.org/project/multipledispatch/)
2. `@singledispatch` from `functools`
3. `@overload` from `typing`
4. Create our own decorator

**Reply by said-hadjout** [reply]
It is preferable to use something that will continue to be maintained, so maybe the second or the third

**Reply by EmilyBourne** [reply]
I agree. Anything built into Python is preferable

**Reply by said-hadjout** [reply]
`@overload` is generally used with mypy to do some type checking

**Reply by EmilyBourne** [reply]
Do you know if mypy can handle `@singledispatch`? The types are also specified in a mypy friendly way so as long as mypy is aware of `@singledispatch` this should also be true there

**Reply by EmilyBourne** [reply]
> `@overload` is generally used with mypy to do some type checking

If generally = commonly, then we may want to support this syntax as an alternative to `@template`

**Reply by said-hadjout** [reply]
![image](https://github.com/pyccel/pyccel/assets/150311653/d32b7165-38cf-46bc-b98c-dcd77e3ee39e)
`overload` does nothing in Python it is only used in stub files

**Reply by said-hadjout** [reply]
then probably `singledispatch` is the best solution if you want it to behave properly

**Reply by said-hadjout** [reply]
I'm unsure if we can replace `@template`, all the proposed solutions have different behaviors. You want to use `@singledispatch`  if you have multiple function definitions. With the way we use `@template` we only need one function definition.

**Reply by EmilyBourne** [reply]
Interesting, so we should probably use this when we add stub files

**Reply by said-hadjout** [reply]
yes, so we are left with `@singledistpatch` but we need to make sure that it can understand our datatypes

**Reply by EmilyBourne** [reply]
I agree. I prefer `@template` as it is usually easier to shorter code. However I'm not sure how this works in stub files. Can we still use imported homemade decorators or do we need to write each function type explicitly. This could be written with `@overload` and would be easy if the file is auto-generated

**Reply by EmilyBourne** [reply]
> yes, so we are left with @singledispatch but we need to make sure that it can understand our datatypes

Do you foresee any particular problems (except for the fact that it obviously should not be combined with `@template`)? I believe our datatypes should be compatible with mypy

**Reply by said-hadjout** [reply]
`@template` will have no issue in the stub files, stub files are only used to store signatures, they are not executed, the issue will probably be when we run them using mypy, but I don't think people care about mypy

**Reply by said-hadjout** [reply]
this is the pyi of `singledispatch`  https://github.com/python/typeshed/blob/2e2c6fca8b792b0ae49c0c6079dc57cd4c6574d9/stubs/singledispatch/singledispatch.pyi, we can use it

**Reply by said-hadjout** [reply]
If you want, we can always implement it ourselves by doing something like
```python
def overload(f):
    oldfunc = globals()[f.__name__]
    param1 = f.__code__.co_varnames[0]
    type1 = f.__annotations__[param1]
    def impl(*args, **kwargs):
        if param1 in kwargs:
            arg1 = kwargs[param1]
        else:
            arg1 = args[0]
        if isinstance(arg1, type1):
            return f(*args, **kwargs)
        else:
            return oldfunc(*args, **kwargs)
    impl.__name__ = f.__name__
    return impl

import math
import numpy

@overload
def sin(x: float):
    print('float')
    return math.sin(x)

@overload
def sin(x: numpy.ndarray):
    print('numpy array')
    return numpy.sin(x)

sin(1.2)
# float
# 0.9320390859672263

sin(numpy.arange(10))
# numpy array
# array([ 0.        ,  0.84147098,  0.90929743,  0.14112001, -0.7568025 ,
#        -0.95892427, -0.2794155 ,  0.6569866 ,  0.98935825,  0.41211849])
```

**Reply by EmilyBourne** [reply]
But then we have to maintain it. If there is an existing Python solution I would prefer this so users can avoid importing from Pyccel where possible should they want to keep their code isolated

**Reply by said-hadjout** [reply]
okay

**Reply by said-hadjout** [reply]
In this case `@singledispatch` is the only solution

**Reply by EmilyBourne** [reply]
But we still have the problem for specifying a `numpy.ndarray`

**Reply by EmilyBourne** [reply]
It's annoying that NumPy still haven't concluded about the right way to manage this

**Reply by said-hadjout** [reply]
yes, it is a problem. if `singledispatch` does not work with array,  then the only solution that I see for the moment is to create our own decorator, and maybe change it in the future if we see the need to

**Reply by EmilyBourne** [reply]
Maybe we can try and discuss this in tomorrow's Pyccel meeting?
@said-hadjout are you on the discord?

**Reply by said-hadjout** [reply]
Yes, discussing that in the meeting is better, but no, I'm not on discord

**Reply by EmilyBourne** [reply]
The link is here: [Pyccel Discord Server](https://discord.gg/2Q6hwjfFVb).

It looks like someone tried to add you at some point but you never accepted the invite. We usually share the links for meetings on there

**Reply by said-hadjout** [reply]
I'm in
